keystore/v2/keystore/filesystem/keyStore.go

```diff
@@ -0,0 +1,206 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filesystem
+
+import (
+	"errors"
+	"runtime"
+	"time"
+
+	"github.com/cossacklabs/acra/keystore/v2/keystore/api"
+	"github.com/cossacklabs/acra/keystore/v2/keystore/asn1"
+	"github.com/cossacklabs/acra/keystore/v2/keystore/crypto"
+	"github.com/cossacklabs/acra/keystore/v2/keystore/filesystem/backend"
+	"github.com/cossacklabs/acra/keystore/v2/keystore/signature"
+	log "github.com/sirupsen/logrus"
+)
+
+const serviceName = "keystore"
+
+// KeyStore is a filesystem-like key store which keeps key rings in files.
+//
+// What exactly is the underlying filesystem is somewhat flexible and controlled by filesystem.Backend.
+// Normally this is an actual filesystem but there are alternative implementations.
+type KeyStore struct {
+	encryptor crypto.KeyStoreSuite
+	notary    *signature.Notary
+	log       *log.Entry
+	fs        backend.Backend
+}
+
+// OpenDirectory opens a read-only key store located in given directory.
+func OpenDirectory(rootDir string, cryptosuite crypto.KeyStoreSuite) (api.KeyStore, error) {
+	backend, err := backend.OpenDirectoryBackend(rootDir)
+	if err != nil {
+		return nil, err
+	}
+	return CustomKeyStore(backend, cryptosuite)
+}
+
+// OpenDirectoryRW opens a key store located in given directory.
+// If the directory does not exist it will be created.
+func OpenDirectoryRW(rootDir string, cryptosuite crypto.KeyStoreSuite) (api.MutableKeyStore, error) {
+	backend, err := backend.CreateDirectoryBackend(rootDir)
+	if err != nil {
+		return nil, err
+	}
+	return CustomKeyStore(backend, cryptosuite)
+}
+
+// NewInMemory returns a new, empty in-memory key store.
+// This is mostly useful for testing.
+func NewInMemory(cryptosuite crypto.KeyStoreSuite) (api.MutableKeyStore, error) {
+	return CustomKeyStore(backend.NewInMemory(), cryptosuite)
+}
+
+// CustomKeyStore returns a configurable filesystem-based key store.
+// This constructor is useful if you want to provide a custom filesystem backend.
+//
+// The backend will be closed when this key store is closed,
+// so a backend instance generally cannot be shared between key stores.
+func CustomKeyStore(backend backend.Backend, cryptosuite crypto.KeyStoreSuite) (api.MutableKeyStore, error) {
+	notary, err := signature.NewNotary(cryptosuite)
+	if err != nil {
+		return nil, err
+	}
+	keystore := &KeyStore{
+		encryptor: cryptosuite,
+		notary:    notary,
+		fs:        backend,
+		log: log.WithFields(log.Fields{
+			"service": serviceName,
+		}),
+	}
+	runtime.SetFinalizer(keystore, (*KeyStore).finalize)
```

Comment:
> how you found this method and decided to use?)

Well, I know about garbage collection probably more than I would like to know, so... :)

It’s probably overkill, but it’s a safety measure nevertheless. The motivation was to ensure that we release the file locks on the key store and close the backend connections in case something somewhere panics but we recover (and in case developers forget to write `defer keystore.Close()`).

Actually, `os.File` definitely has a finalizer and various collection handles tend to have them as well (at least indirectly via some `net.TCPConn` instance they wrap), so the OS resources will be released eventually in case they leak. However, this works only for filesystem locks (they are released when the last file descriptor for the lock file is closed). This may not be the case with distributed locks implemented on top of some key-value storage. In this case we will need to perform more work than simply close a socket.

Though, key stores tend to be allocated in `main()` and closed only before exiting the process so we’d close the file descriptor anyway. And in case of distributed locks we’d still need to make them work with timed leases or something to protect against connectivity losses, so they too will be eventually released even if the holder does not do it explicitly.

So this finalizer is here mostly for the peace of mind, and for that sake of that `log.Warn()` in case a key store is not allocated in `main()` _and_ leaked without being closed. (Go does not run pending finalizers when the process exits, so if you leak a key store in main() then you won’t see that warning.)